src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs

```diff
@@ -16,65 +16,78 @@ namespace Microsoft.CodeAnalysis
     internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
     {
         /// <summary>
-        /// The intended size of the <see cref="Sha1Hash"/> structure. Some runtime environments are known to deviate
-        /// from this size, so the implementation code should remain functionally correct even when the structure size
-        /// is greater than this value.
+        /// The intended size of the <see cref="HashData"/> structure. 
         /// </summary>
-        /// <seealso href="https://bugzilla.xamarin.com/show_bug.cgi?id=60298">LayoutKind.Explicit, Size = 12 ignored with 64bit alignment</seealso>
-        /// <seealso href="https://github.com/dotnet/roslyn/issues/23722">Checksum throws on Mono 64-bit</seealso>
-        private const int Sha1HashSize = 20;
+        private const int HashSize = 20;
 
-        public static readonly Checksum Null = new Checksum(Array.Empty<byte>());
+        public static readonly Checksum Null = new Checksum(default);
 
-        private Sha1Hash _checkSum;
+        private HashData _checksum;
 
-        public unsafe Checksum(byte[] checksum)
+        public static unsafe Checksum From(byte[] checksum, bool truncate = false)
         {
+            Validate(checksum.Length, truncate);
+
             if (checksum.Length == 0)
             {
-                _checkSum = default;
-                return;
-            }
-            else if (checksum.Length != Sha1HashSize)
-            {
-                throw new ArgumentException($"{nameof(checksum)} must be a SHA-1 hash", nameof(checksum));
+                return Null;
             }
 
             fixed (byte* data = checksum)
             {
-                // Avoid a direct dereferencing assignment since sizeof(Sha1Hash) may be greater than Sha1HashSize.
-                _checkSum = Sha1Hash.FromPointer((Sha1Hash*)data);
+                // Avoid a direct dereferencing assignment since sizeof(HashData) may be greater than Sha1HashSize.
+                return new Checksum(HashData.FromPointer((HashData*)data));
             }
         }
 
-        public unsafe Checksum(ImmutableArray<byte> checksum)
+        public static unsafe Checksum From(ImmutableArray<byte> checksum, bool truncate = false)
```

Comment:
What is using this and _isn't_ truncating? Don't we use the same hash algorithm everywhere?

Reply:
we have something like SourceText that return ImmutableArray<byte> rather than byte[].

when we serialize and deserialized checksum, we don't truncate.

Reply:
We don't truncate because we're actually using the full bits, or we're not truncating because the checksum is already at the right size?

Reply:
it is already in right size.

Reply:
I don't get it. Why don't we only throw if we get less than 20 bytes and implicitly truncate when we get more?

---
In reply to: [257065181](https://github.com/dotnet/roslyn/pull/33363#discussion_r257065181) [](ancestors = 257065181)

Reply:
I wanted to make truncate explicit. but sure. I can make it implicit if that make people feel better.